### PR TITLE
Upgrade to VS2017 and new csproj format

### PR DESCRIPTION
**Pull request by gpetrou**

### CODE REVIEW

File: Tests/Metadata/Metadata.csproj [hunk 1]
```diff
@@ -12,39 +12,8 @@
...
-    <ProjectReference Include="..\..\Src\Microsoft.Dynamic\Microsoft.Dynamic.csproj">
-      <Project>{EB66B766-6354-4208-A3D4-AACBDCB5C3B3}</Project>
-      <Name>Microsoft.Dynamic</Name>
-    </ProjectReference>
-    <ProjectReference Include="..\..\Src\Microsoft.Scripting.Metadata\Microsoft.Scripting.Metadata.csproj">
-      <Project>{ACDD9B9E-8FE6-439C-9521-1CCBA47F6143}</Project>
-      <Name>Microsoft.Scripting.Metadata</Name>
-    </ProjectReference>
+    <ProjectReference Include="..\..\Src\Microsoft.Scripting.Metadata\Microsoft.Scripting.Metadata.csproj" />
+    <ProjectReference Include="..\..\Src\Microsoft.Scripting\Microsoft.Scripting.csproj" />
```
Comment (slozier) [inline Tests/Metadata/Metadata.csproj]: This one should point to Microsoft.Dynamic.

Comment (gpetrou) [inline Tests/Metadata/Metadata.csproj]: Ooops, yes, I forgot to revert this. The problem I have is something else though with the ReflectionUtils.

Comment (slozier) [inline Tests/Metadata/Metadata.csproj]: In the AssemblyInfo files for Microsoft.Scripting and Microsoft.Dynamic there are some `InternalsVisibleTo("TestInternalDLR, ...")` which may be what's causing the issues. You can comment those out.

File: Src/Microsoft.Scripting/Microsoft.Scripting.csproj [hunk 2]
```diff
@@ -11,133 +11,9 @@
...
-  </PropertyGroup>
-  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'v4Release|AnyCPU' ">
-  </PropertyGroup>
-  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
-  </PropertyGroup>
-  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
-  </PropertyGroup>
+  </PropertyGroup>  
   <ItemGroup>
-    <Compile Include="..\AssemblyInfo.cs">
```
Comment (slozier) [inline Src/Microsoft.Scripting/Microsoft.Scripting.csproj]: I think we'll want to keep this one (here and also in the other csproj files). I probably inadvertently removed it from my original PR. Note that you make have to change the Link to something else (e.g. Properties\AssemblyInfo.Generated.cs) if it causes conflicts with an existing AssemblyInfo.cs file.

File: Build/Common.proj [hunk 3]
```diff
@@ -49,7 +55,6 @@
     <BuildFlavour>Debug</BuildFlavour>
     <BuildFlavour Condition="'$(Configuration)' == 'FxCop'">Release</BuildFlavour>
     <BuildFlavour Condition="'$(Configuration)' == 'Release'">Release</BuildFlavour>
-    <BuildFlavour Condition="'$(Configuration)' == 'v4Release'">Release</BuildFlavour>
```
Comment (slide) [inline Build/Common.proj]: Why was this removed?

File: Tests/Metadata/Metadata.csproj (hunk 1 above)
Comment (slide) [inline Tests/Metadata/Metadata.csproj]: Those are needed for tests though, I believe.

Comment (slozier) [inline Tests/Metadata/Metadata.csproj]: You're right, although it looks like we only use the internals of Microsoft.Scripting so we can exclude it from Microsoft.Dynamic.

**Comment by slide** [comment]
It would be nice if .travis.yml would use the makefile instead.

**Comment by slide** [comment]
Does this change mean that you MUST have VS2017 in order to build?

**Comment by slozier** [comment]
If you want to use VS then yes, it won't work with VS2015.

**Comment by slide** [comment]
I'm not sure I have access to VS2017

**Comment by gpetrou** [comment]
If I am not wrong the new csproj format is compatible only with VS2017.
So, yes, you will need to get VS2017.
I am not sure about the makefile in travis. Maybe this should be a separate issue/PR?

**Comment by slozier** [comment]
Looks like there's still some stuff ending up in the bin/Release folder. For example Microsoft.Scripting.AspNet which I hadn't converted to the new csproj because I have no idea what it's used for.

**Comment by slide** [comment]
It's used for doing ASP.NET Webform (I believe) in IronPython. I think we can probably get rid of it since Webforms is not really used much anymore.

**Comment by gpetrou** [comment]
I was planning to add a separate PR for the rest of the cspoj files.
Some of them don't even seem to be part of the sln.

File: Tests/ClrAssembly/ClrAssembly.csproj [hunk 4]
```diff
@@ -1,25 +1,20 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
-<Project ToolsVersion="4.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+﻿<Project Sdk="Microsoft.NET.Sdk">
   <Import Project="$(ProjectDir)..\..\Build\Common.proj" />
   <PropertyGroup>
-    <SchemaVersion>2.0</SchemaVersion>
+    <TargetFrameworks>net40;net45</TargetFrameworks>
+    <PlatformTarget>anycpu</PlatformTarget>
```
Comment (slozier) [inline Tests/ClrAssembly/ClrAssembly.csproj]: Can you get rid of all those extra properties that were added?

**Comment by slozier** [comment]
The reason it's failing on Travis is that it's using Build.proj which will try to build Release and v4Release (and v4Release doesn't exist anymore). If you remove the `<ReferencedPlatform Include="net40">...` it should work. We'll probably need to do some more cleanup of Build.proj.

File: Tests/ClrAssembly/ClrAssembly.csproj (hunk 4 above)
Comment (gpetrou) [inline Tests/ClrAssembly/ClrAssembly.csproj]: Sorry, I am still experimenting please ignore for now.

**Comment by slozier** [comment]
You can add `<EnableDefaultCompileItems>false</EnableDefaultCompileItems>` to ClrAssembly.csproj to resolve the compile issues (you'll need to re-add dummy.cs to get rid of the warning).

**Comment by gpetrou** [comment]
Indeed, there is some information in https://docs.microsoft.com/en-us/dotnet/core/tools/csproj

File: Src/Microsoft.Dynamic/Microsoft.Dynamic.csproj [hunk 5]
```diff
@@ -12,433 +12,16 @@
...
-    <Reference Include="System.Configuration" />
-    <Reference Include="System.Runtime.Remoting" />
-  </ItemGroup>
-  <ItemGroup Condition="'$(ReferencedPlatform)' == 'net40'">
-    <Reference Include="System.Xaml" />
-  </ItemGroup>
-  <ItemGroup Condition="'$(ReferencedPlatform)' == 'net45'">
-    <Reference Include="System.Xaml" />
-  </ItemGroup>
+    <ProjectReference Include="..\Microsoft.Scripting.Metadata\Microsoft.Scripting.Metadata.csproj" />
```
Comment (slozier) [inline Src/Microsoft.Dynamic/Microsoft.Dynamic.csproj]: Can you add a `Condition=" $(Features.Contains('FEATURE_METADATA_READER')) "` to this reference?

**Comment by slozier** [comment]
I tried building locally and it looks like ClrAssembly no longer produces the test DLLs?

File: Tests/ClrAssembly/ClrAssembly.csproj [hunk 6]
```diff
@@ -99,15 +93,11 @@
   </Target>
   <Target Name="BeforeBuild" DependsOnTargets="LoadOrder;ReachType;CSManyArea;VBManyArea" />
   <ItemGroup>
-    <Compile Include="dummy.cs" />
```
Comment (slozier) [inline Tests/ClrAssembly/ClrAssembly.csproj]: Can you keep this one in the project (to eliminate the CS2008 build warning).

File: Src/Microsoft.Dynamic/Microsoft.Dynamic.csproj (hunk 5 above)
Comment (gpetrou) [inline Src/Microsoft.Dynamic/Microsoft.Dynamic.csproj]: Yes, sure. Is it causing a problem?

Comment (slozier) [inline Src/Microsoft.Dynamic/Microsoft.Dynamic.csproj]: Not at the moment, but it's only used when that condition is met and will need to be excluded for .NET Core.

File: Src/Microsoft.Dynamic/Microsoft.Dynamic.csproj [hunk 7]
```diff
@@ -23,5 +23,9 @@
     <Reference Include="System.Configuration" Condition=" $(Features.Contains('FEATURE_CONFIGURATION')) " />
     <Reference Include="System.Runtime.Remoting" Condition=" $(Features.Contains('FEATURE_REMOTING')) " />
     <Reference Include="System.Xaml" Condition=" $(Features.Contains('FEATURE_WPF')) " />
-   </ItemGroup>
+  </ItemGroup>
+  <Import Project="$(BeforeTargetFiles)" />
+  <Target Name="BeforeBuild" DependsOnTargets="$(BeforeTargets);GenerateCurrentVersion" />
```
Comment (slozier) [inline Src/Microsoft.Dynamic/Microsoft.Dynamic.csproj]: I think these won't work properly (same issue as before, the target name BeforeBuild doesn't seem to work with the new project files).

File: Src/Microsoft.Dynamic/Microsoft.Dynamic.csproj [hunk 8]
```diff
@@ -24,8 +24,4 @@
     <Reference Include="System.Runtime.Remoting" Condition=" $(Features.Contains('FEATURE_REMOTING')) " />
     <Reference Include="System.Xaml" Condition=" $(Features.Contains('FEATURE_WPF')) " />
   </ItemGroup>
-  <Import Project="$(BeforeTargetFiles)" />
-  <Target Name="BeforeBuild" DependsOnTargets="$(BeforeTargets);GenerateCurrentVersion" />
```
Comment (slozier) [inline Src/Microsoft.Dynamic/Microsoft.Dynamic.csproj]: We're still going to need these. We just have to change them to something like `<Target Name="BeforeBuildStart" BeforeTargets="BeforeBuild"`.

Comment (gpetrou) [inline Src/Microsoft.Dynamic/Microsoft.Dynamic.csproj]: Hmm, I did the change but still get some errors on a clean build.

**Comment by slozier** [comment]
Looks like it's still not generating the CurrentVersion files. We'll probably have to tweak it some more to get the target to run. I'm gonna have to do some reading on the subject...

**Comment by gpetrou** [comment]
I tried to do it like https://stackoverflow.com/questions/43921992/how-can-i-use-beforebuild-and-afterbuild-targets-with-visual-studio-2017 but no luck yet.

**Comment by slozier** [comment]
I ran a few tests. Looks like the files are getting generated they're just not picked up automatically. Maybe BeforeTargets needs to be something other than BeforeBuild? Maybe there's a target that occurs before file detection?

I found a workaround but it's ugly so I'm open to suggestions. I added the following after the BeforeBuildStarts targets:
```
  <ItemGroup>
    <!-- if the file does not exist it's not picked up automatically -->
    <Compile Remove="Properties\CurrentVersion.Generated.cs" />
    <Compile Include="Properties\CurrentVersion.Generated.cs" />
  </ItemGroup>
```

**Comment by gpetrou** [comment]
I had a look around, but could not find another solution, so I used your suggestion for now.

File: Src/Microsoft.Scripting/Microsoft.Scripting.csproj [hunk 9]
```diff
@@ -1,8 +1,9 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
-<Project ToolsVersion="4.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+﻿<Project>
+  <Import Project="Sdk.props" Sdk="Microsoft.NET.Sdk" />
```
Comment (slozier) [inline Src/Microsoft.Scripting/Microsoft.Scripting.csproj]: You're doing the separate target/props import instead of using sdk in the project tag so that the BeforeBuild/AfterBuild targets work? Perhaps a comment to that effect in the project files would be useful.

**Comment by gpetrou** [comment]
I can do a PR for the rest of the csproj files once this is merged.
Would it be better to convert the make.cmd to a PowerShell script?
Personally, I find these files harder to work with unless they are doing very simple tasks.

**Comment by slide** [comment]
I would prefer something like fake or psake if possible.

**Comment by slide** [comment]
Actually, I would prefer we move everything to msbuild so we could use the same commands on both windows and Linux, e.g., msbuild /t:test-all
